src/plugins/application_config/server/plugin.ts

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { Observable } from 'rxjs';
+import { first } from 'rxjs/operators';
+
+import {
+  PluginInitializerContext,
+  CoreSetup,
+  CoreStart,
+  Plugin,
+  Logger,
+  IScopedClusterClient,
+  SharedGlobalConfig,
+} from '../../../core/server';
+
+import {
+  ApplicationConfigPluginSetup,
+  ApplicationConfigPluginStart,
+  ConfigurationClient,
+} from './types';
+import { defineRoutes } from './routes';
+import { OpenSearchConfigurationClient } from './opensearch_config_client';
+
+export class ApplicationConfigPlugin
+  implements Plugin<ApplicationConfigPluginSetup, ApplicationConfigPluginStart> {
+  private readonly logger: Logger;
+  private readonly config$: Observable<SharedGlobalConfig>;
+
+  private configurationClient: ConfigurationClient;
+  private configurationIndexName: string;
+
+  constructor(initializerContext: PluginInitializerContext) {
+    this.logger = initializerContext.logger.get();
+    this.config$ = initializerContext.config.legacy.globalConfig$;
+    this.configurationIndexName = '';
+  }
+
+  private setConfigurationClient(inputConfigurationClient: ConfigurationClient) {
+    this.configurationClient = inputConfigurationClient;
+  }
+
+  private getConfigurationClient(inputOpenSearchClient: IScopedClusterClient) {
+    if (this.configurationClient) {
+      return this.configurationClient;
+    }
+
+    const openSearchConfigurationClient = new OpenSearchConfigurationClient(
+      inputOpenSearchClient,
+      this.configurationIndexName,
+      this.logger
+    );
+
+    this.setConfigurationClient(openSearchConfigurationClient);
+    return this.configurationClient;
+  }
+
+  public async setup(core: CoreSetup) {
+    this.logger.debug('applicationConfig: Setup');
+    const router = core.http.createRouter();
+
+    const config = await this.config$.pipe(first()).toPromise();
+
+    this.configurationIndexName = config.opensearchDashboards.config_index;
+
+    // Register server side APIs
+    defineRoutes(router, this.getConfigurationClient.bind(this), this.logger);
```

Comment:
I thought the client is initiated already and we should be able to use the object. Are we not able to get the open search client object when the plugin is initiated?

Reply:
We need a scoped opensearch client which is available from a request context. This is for user permission against the index through calling `asCurrentUser` for update and delete. As a contrast, we are calling `asInternalUser` for read. Thus it is not yet there when plugin is initialized. 

It will need the caller to pass in their scoped client.